cpp/examples/arrow/rapidjson_row_converter.cc

```diff
@@ -0,0 +1,590 @@
+//
+// We use the following helpers and patterns here:
+//  * arrow::ToRowConverter and arrow::FromRowConverter, which provide additional
+//    conversion methods given a basic converter
+//  * arrow::VisitArrayInline and arrow::VisitTypeInline for implementing a visitor
+//    pattern with Arrow to handle different array types
+//  * arrow::enable_if_primitive_ctype to create a template method that handles
+//  * conversion for Arrow types that have corresponding C types (bool, integer,
+//    float).
+
+rapidjson::Value kNullJsonSingleton = rapidjson::Value();
+
+/// \brief Builder that holds state for a single conversion.
+///
+/// Implements Visit() methods for each type of Arrow Array that set the values
+/// of the corresponding fields in each row.
+class RowBatchBuilder {
+ public:
+  explicit RowBatchBuilder(int64_t num_rows) : field_(nullptr) {
+    // Reserve all of the space required up-front to avoid unnecessary resizing
+    rows_.reserve(num_rows);
+
+    for (int64_t i = 0; i < num_rows; ++i) {
+      rows_.push_back(rapidjson::Document());
+      rows_[i].SetObject();
+    }
+  }
+
+  /// \brief Set which field to convert.
+  void SetField(std::shared_ptr<arrow::Field> field) { field_ = std::move(field); }
+
+  /// \brief Retrieve converted rows from builder.
+  std::vector<rapidjson::Document> Rows() && { return std::move(rows_); }
+
+  // Default implementation
+  arrow::Status Visit(const arrow::Array& array) {
+    return arrow::Status::NotImplemented(
+        "Can not convert to json document for array of type ", array.type()->ToString());
+  }
+
+  // Handles booleans, integers, floats
+  template <typename ArrayType, typename DataClass = typename ArrayType::TypeClass>
+  arrow::enable_if_primitive_ctype<DataClass, arrow::Status> Visit(
+      const ArrayType& array) {
+    for (int64_t i = 0; i < array.length(); ++i) {
+      if (!array.IsNull(i)) {
+        rapidjson::Value str_key(field_->name().c_str(), rows_[i].GetAllocator());
+        rows_[i].AddMember(str_key, array.Value(i), rows_[i].GetAllocator());
+      }
+    }
+    return arrow::Status::OK();
+  }
+
+  arrow::Status Visit(const arrow::StringArray& array) {
+    for (int64_t i = 0; i < array.length(); ++i) {
+      if (!array.IsNull(i)) {
+        rapidjson::Value str_key(field_->name().c_str(), rows_[i].GetAllocator());
+        // StringArray.Value returns a string view
+        auto value_view = array.Value(i);
+        rapidjson::Value value;
+        value.SetString(value_view.data(),
+                        static_cast<rapidjson::SizeType>(value_view.size()),
+                        rows_[i].GetAllocator());
+        rows_[i].AddMember(str_key, value, rows_[i].GetAllocator());
+      }
+    }
+    return arrow::Status::OK();
+  }
+
+  arrow::Status Visit(const arrow::StructArray& array) {
+    const arrow::StructType* type = array.struct_type();
+
+    RowBatchBuilder child_builder(rows_.size());
```

Comment:
should this be based on array.size()?

Reply:
or is this saying only extract n_rows from the struct array?

Reply:
The expectation is that the class is constructed with the correct number of rows, so `rows_.size() == array.length()`. But I think it's worth adding an `assert` here to make that explicit.